From d84afefab75a67946e1c3aa9585dd6b594e876d3 Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zahari.doychev@intel.com>
Date: Mon, 21 Sep 2020 15:52:10 +0200
Subject: [PATCH] pon_qos: fix ctp handling for mirred offloads

The driver was not setting the subif groupid when needed and was using the DP
subif instead of the subif groupid. This was causing the offload to failed
when used on VUNI0_0.
---
 drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c |  8 ++++----
 drivers/net/datapath/pon_qos/pon_qos_tc_pce.c    | 22 +++++++++++++---------
 2 files changed, 17 insertions(+), 13 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c b/drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c
index 1a7a0b13d076..56ec421c2b34 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c
@@ -116,10 +116,10 @@ static int pon_qos_vlan_to_pce_single(struct net_device *dev,
 	rule->action.ePortFilterType_Action = 1;
 
 	if (flt->drop) {
-		pr_debug("%s: Mirring VLAN forward\n", __func__);
+		pr_debug("%s: Mirring VLAN drop\n", __func__);
 		flt->pce_type = PCE_MIRR_VLAN_DROP;
 	} else {
-		pr_debug("%s: Mirring VLAN drop\n", __func__);
+		pr_debug("%s: Mirring VLAN forward\n", __func__);
 		flt->pce_type = PCE_MIRR_VLAN_FWD;
 	}
 
@@ -197,13 +197,13 @@ static int pon_qos_tc_to_pce(struct net_device *dev,
 	}
 
 	netdev_dbg(dev, "%s: using port: %d subif: %d\n",
-		   __func__, dp_subif.port_id, dp_subif.subif);
+		   __func__, dp_subif.port_id, dp_subif.subif_groupid);
 
 	pce_rule->pattern.bEnable = 1;
 	pce_rule->pattern.bPortIdEnable = 1;
 	pce_rule->pattern.nPortId = dp_subif.port_id;
 	pce_rule->pattern.bSubIfIdEnable = 1;
-	pce_rule->pattern.nSubIfId = dp_subif.subif;
+	pce_rule->pattern.nSubIfId = dp_subif.subif_groupid;
 
 	if (flt->proto == ETH_P_ALL) {
 		ret = pon_qos_all_to_pce(dev, flt, pce_rule);
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c b/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
index 7f51933a9d59..056dfcf9255c 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
@@ -144,8 +144,9 @@ static int pce_free_index_get(u32 *index,
 		ret = gsw_handle->gsw_tflow_ops.TFLOW_PceRuleRead(gsw_handle,
 								  &pce_rule);
 		if (ret != GSW_statusOk) {
-			pr_err("%s: TFLOW_PceRuleRead failed, nIndex: %u, ret: %d\n",
-			       __func__, pce_rule.pattern.nIndex, ret);
+			pr_err("%s: lpid/gpid: %u/%u failed, idx: %u, ret: %d\n",
+			       __func__, portid, subif,
+			       pce_rule.pattern.nIndex, ret);
 			return ret;
 		}
 
@@ -206,10 +207,11 @@ static int get_subif_data(struct net_device *dev,
 		ret = dp_get_netif_subifid(dev, NULL, NULL, NULL, &dp_subif, 0);
 		if (ret != DP_SUCCESS)
 			return -ENODEV;
-		pr_debug("%s: portid: %u, subif: %u\n", __func__,
-			 dp_subif.port_id, dp_subif.subif);
+
+		pr_debug("%s: portid: %u, subif: %u\n",
+			 __func__, dp_subif.port_id, dp_subif.subif_groupid);
 		*portid = dp_subif.port_id;
-		*subif = dp_subif.subif;
+		*subif = dp_subif.subif_groupid;
 		break;
 	default:
 		pr_err("%s: PCE type unknown\n", __func__);
@@ -250,6 +252,7 @@ static int pce_block_create(int portid, int subif,
 			 __func__, alloc.num_of_rules, alloc.blockid);
 
 		ctp.nLogicalPortId = portid;
+		ctp.nSubIfIdGroup = subif;
 		ctp.nFirstFlowEntryIndex = alloc.blockid;
 		ctp.eMask |= GSW_CTP_PORT_CONFIG_FLOW_ENTRY;
 
@@ -284,7 +287,7 @@ static int pce_block_create(int portid, int subif,
 
 	*block = pce_block;
 
-	pr_debug("%s: PCE block created\n", __func__);
+	pr_debug("%s: PCE block created for subif: %#x\n", __func__, subif);
 	return 0;
 }
 
@@ -358,7 +361,7 @@ int pon_qos_pce_rule_create(struct net_device *dev,
 	block->reserved[index] = PON_QOS_PCE_RESERVED;
 	block->reservation_cnt++;
 
-	pr_debug("%s: PCE rule created\n", __func__);
+	pr_debug("%s: PCE rule created for lpid/gpid %u/%u\n", __func__, portid, subif);
 	return 0;
 
 error_block:
@@ -382,6 +385,7 @@ static int pce_block_delete(int portid, int subif,
 	}
 
 	ctp.nLogicalPortId = portid;
+	ctp.nSubIfIdGroup = subif;
 	ctp.nFirstFlowEntryIndex = 0xFFFFFFFF;
 	ctp.eMask |= GSW_CTP_PORT_CONFIG_FLOW_ENTRY;
 
@@ -400,8 +404,8 @@ static int pce_block_delete(int portid, int subif,
 		return ret;
 	}
 
-	pr_debug("%s: CTP_PortConfigSet success, nLogicalPortId: %u, nFirstFlowEntryIndex: %u\n",
-		 __func__, ctp.nLogicalPortId, ctp.nFirstFlowEntryIndex);
+	pr_debug("%s: CTP_PortConfigSet success, lpid/gpid: %u/%u, nFirstFlowEntryIndex: %u\n",
+		 __func__, portid, subif, ctp.nFirstFlowEntryIndex);
 	alloc.blockid = block->blockid;
 
 	ret = gsw_handle->gsw_tflow_ops.TFLOW_PceRuleFree(gsw_handle, &alloc);
